src/ast/mod.rs

```diff
@@ -147,15 +150,121 @@ impl fmt::Display for Ident {
 }
 
 /// A name of a table, view, custom type, etc., possibly multi-part, i.e. db.schema.obj
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+///
+/// # Semantic Equality
+///
+/// This implements equals and hash in terms of the ident *values*, not
+/// including quotes.
+///
+/// ```
+/// use sqlparser::ast::{Ident, ObjectName};
+///
+/// let with = Ident::with_quote('"', "one");
+/// let without = Ident::new("one");
+/// assert_eq!(with.to_string(), "\"one\"");
+/// assert_eq!(without.to_string(), "one");
+/// assert_eq!(ObjectName(vec![with]), ObjectName(vec![without]));
+/// ```
+///
+/// If you need to compare based on quotes see the
+/// [`ObjectName::quoted_string`] method.
+#[derive(Debug, Clone, Eq)]
 pub struct ObjectName(pub Vec<Ident>);
 
+impl PartialEq for ObjectName {
+    fn eq(&self, rhs: &ObjectName) -> bool {
+        if self.0.len() != rhs.0.len() {
+            return false;
+        }
+
+        for (l, r) in self.0.iter().zip(rhs.0.iter()) {
+            if l.value != r.value {
+                return false;
+            }
+        }
+
+        true
+    }
+}
+
+impl Hash for ObjectName {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        for ident in &self.0 {
+            state.write(ident.value.as_bytes());
+        }
+    }
+}
+
 impl fmt::Display for ObjectName {
+    /// Construct an object name including quotes around all internal idents
+    ///
+    /// Uses the format of [`ObjectName::quoted_string`].
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         write!(f, "{}", display_separated(&self.0, "."))
     }
 }
 
+impl ObjectName {
+    /// Construct an object name with no quotes around the internal items
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use sqlparser::ast::{Ident, ObjectName};
+    ///
+    /// let on = ObjectName(vec![Ident::new("one"), Ident::with_quote('"', "two sp")]);
+    /// assert_eq!(on.unquoted_string(), "one.two sp");
+    /// ```
+    pub fn unquoted_string(&self) -> String {
+        let ident_count = self.0.len();
+        match ident_count {
+            0 => unreachable!("Object created without name"),
+            1 => self.0[0].value.clone(),
+            _ => {
+                let cap: usize = self.0.iter().map(|i| i.value.len() + 1).sum();
+                let mut buf = String::with_capacity(cap - 1);
+                for (i, ident) in self.0.iter().enumerate() {
+                    buf.push_str(&ident.value);
+                    if i != ident_count - 1 {
+                        buf.push('.');
+                    }
+                }
+                buf
+            }
+        }
+    }
```

Comment:
I'm not sure I understand the purpose of this method. It seems dangerous in that

```rust
let on1 = ObjectName(vec![Ident::new("a"), Ident::with_quote('"', "b.c")]);
let on2 = ObjectName(vec![Ident::new("a"), Ident::new("b"), Ident::new("c")])
```

will have the same unquoted representation, but they are definitely not the same ident. At the very least I don't think it belongs in the parser.

Reply:
ah yeah I'll extract this and do the work where we care about semantics.